src/debug.cc

```diff
@@ -70,19 +99,271 @@ class DebugFile
+class DebugModule
+{
+public:
+    DebugModule();
+
+    // we provide debugging services for the entire duration of the program
+    ~DebugModule() = delete;
+
+    /// \copydoc Debug::SwanSong()
+    void swanSong();
+
+    /// Log the given debugs() message to appropriate channel(s) (eventually).
+    /// Assumes the message has passed the global section/level filter.
+    void log(const DebugMessageHeader &, const std::string &body);
+
+    /// Start using an open cache_log file as the primary debugs() destination.
+    /// Stop using stderr as a cache_log replacement (if we were doing that).
+    void useCacheLog();
+
+    /// Start using stderr as the primary debugs() destination.
+    /// Stop waiting for an open cache_log file (if we were doing that).
+    void banCacheLogUse();
+
+public:
+    CacheLogChannel cacheLogChannel;
+    StderrChannel stderrChannel;
+    SyslogChannel syslogChannel;
+};
+
+/// cache_log file
 /// safe during static initialization, even if it has not been constructed yet
+/// safe during program termination, even if it has been destructed already
 static DebugFile TheLog;
 
 FILE *
 DebugStream() {
-    return TheLog.file();
+    return TheLog.file() ? TheLog.file() : stderr;
+}
+
+/// used for the side effect: fills Debug::Levels with the given level
+static void
+ResetSections(const int level)
+{
+    for (auto &sectionLevel: Debug::Levels)
+        sectionLevel = level;
+}
+
+/* DebugModule */
+
+// Depending on DBG_CRITICAL activity and command line options, this code may
+// run as early as static initialization during program startup or as late as
+// the first debugs(DBG_CRITICAL) call from the main loop.
+DebugModule::DebugModule()
+{
+    // explicit initialization before any use by debugs() calls; see bug #2656
+    tzset();
+
+    (void)std::atexit(&Debug::SwanSong);
+
+    if (!Debug::override_X)
+        ResetSections();
 }
 
 void
-StopUsingDebugLog()
+DebugModule::log(const DebugMessageHeader &header, const std::string &body)
 {
-    TheLog.clear();
+    cacheLogChannel.log(header, body);
+    stderrChannel.log(header, body);
+    syslogChannel.log(header, body);
+}
+
+void
+DebugModule::swanSong()
+{
+    // Switch to stderr to improve our chances to log _early_ debugs(). However,
+    // use existing cache_log and/or stderr levels for post-open/close ones.
+    if (cacheLogChannel.collectingEarlyMessages() && !TheLog.file())
+        banCacheLogUse();
+
+    cacheLogChannel.stopEarlyMessageCollection();
+    stderrChannel.stopEarlyMessageCollection();
+    syslogChannel.stopEarlyMessageCollection();
+
+    // Do not close/destroy channels: While the Debug module is not _guaranteed_
+    // to get control after swanSong(), debugs() calls are still very much
+    // _possible_, and we want to support/log them for as long as we can.
+}
+
+void
+DebugModule::useCacheLog()
+{
+    assert(TheLog.file());
+    stderrChannel.stopCoveringForCacheLog(); // in case it was covering
```

Comment:
I could not find a place where we useCacheLog() after banCacheLogUse(). Is this scenario possible? If not, we don't need to do this, right?

Reply:
I hope it is possible after a reconfiguration that fixes `cache_log /typo/cache.log` spelling, for example.